### ISSUE TITLE
Support keyword values for min- and max-width

### ISSUE THREAD

**Issue by Manishearth** [servo/servo#13821]
min-width, max-width, min-block-size, min-inline-height, (and their max- variants) support [a bunch of keyword values](https://developer.mozilla.org/en-US/docs/Web/CSS/min-width#Formal_syntax).

We should handle these in style, layout, and add stylo glue. At the very least we should have feature-parity with Gecko (which implements these keywords prefixed, but implements them nevertheless)

**Comment by AgostonSzepessy** [comment]
@Manishearth May I try this one?

**Comment by Manishearth** [comment]
Go ahead!

Note that this is a bit tricky; for the logical properties to still work you need to have them share a type, so you have to define a new "predefined" type in values.rs for this. If you think you're up for it go ahead!

**Comment by AgostonSzepessy** [comment]
@Manishearth What do you mean by a "predefined" type? Also, which values.rs are you talking about? Is it `components/style/gecko/values.rs`?

**Comment by Manishearth** [comment]
A predefined type is one in values.rs that can be used with the helpers.predefined_type helper. It automatically gets the gecko glue generated.

**Comment by AgostonSzepessy** [comment]
@Manishearth I'm not sure I understand. `min-width` and `max-width` are already implemented aren't they? On the MDN it says that `min-block-size` and `max-block-size` have the same Formal Syntax. Would I make a predefined type for `min-width` and `max-width` and use that for each one? I also couldn't find `min-inline-height` on this page: https://developer.mozilla.org/en-US/docs/Web/CSS/Reference

**Comment by Manishearth** [comment]
They're implemented as length values, but there are extra keywords in the docs.

the `block-size`/`inline-size` ones are logical properties. That's why this is a bit trickier than usual.

**Comment by AgostonSzepessy** [comment]
@Manishearth So for `block-size / inline-size` I need to get the value of `writing-mode` right? On the MDN pages for these, it says that way each one behaves depends on `writing-mode`. How would I get the value of `writing-mode`? And then would `block-size / inline-size` use the predefined property that I make?

**Comment by Manishearth** [comment]
No, that's handled already. We support all those properties. (see https://github.com/servo/servo/pull/14120)

This bug is about adding keywords to them.

I would recommend you try a different bug; this one isn't very straightforward.

**Comment by AgostonSzepessy** [comment]
@Manishearth Yes, I think you're right. I'll try a different issue.

**Comment by glasserc** [comment]
I'm taking a swing at this, so I thought I'd mention it to prevent getting scooped :)

So far, per discussions on IRC with nox, I have been defining a `specified::MinLength` type and a `computed::MinLength` type. (I will do the same for `MaxLength` once I feel confident that this is the right approach.) Since `MinLength` and `MaxLength` share certain values (`max-content`, `min-content`, `fit-content`, `fill-available`), I extracted a type called `ExtremumLength` which I will share between `MinLength` and `MaxLength`. (I placed it in `values/mod.rs`.)

I am pretty sure I have the right values for `specified::MinLength` (it can be a `Length`, `Percentage`, `Auto`, or an `ExtremumLength`). But in order for it to be animated, it has to be interpolated. Per [min-width on MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/min-width), to interpolate it, if one is a length and the other a percentage, then I need to construct a `calc()` that is the sum of a length and a percentage and animate that. So I think this means that my `computed::MinLength` needs to allow a `Calc` as a possible value; is that correct?

Besides that, I am also getting errors because I don't have a `to_gecko_style_coord` implementation for `computed::MinLength`. I think this is because I wrote in `gecko.mako.rs` that `MinLength` is `impl_style_coord`, but I'm not sure. If I understand this correctly, I need to define a way to map my `MinLength` values into Gecko values -- is that correct? Presumably I can figure out how to handle `Length` and `Percentage` values (since they already exist in other types), but I guess the challenge would be to handle keywords like `fit-content` and `fill-available`. Is mapping these into Gecko values the only thing that this issue requires? Or will I also have to implement Servo-native layout logic for these keywords?

**Comment by Manishearth** [comment]
>  So I think this means that my computed::MinLength needs to allow a Calc as a possible value; is that correct?

Just use `LengthOrPercentage`, LoP already allows for calc values and knows how to interpolate it.

> I need to define a way to map my MinLength values into Gecko values -- is that correct?

Yep. You can no longer use `impl_style_coord`, you will have to manually write glue code (this is done by taking the `impl_style_struct`, adding `min-width` to the skipped longhands, and manually writing set/copy/clone functions). For `LengthOrPercentage` values you can just use `to_gecko_style_coord` and `copy_from` and `from_gecko_style_coord`

In case of keywords, you'll need to map to [these constants](https://dxr.mozilla.org/mozilla-central/rev/0534254e9a40b4bade2577c631fe4cfa0b5db41d/layout/style/nsStyleConsts.h#755)  and use `coord.set_value(CoordDataValue::Enumerated(constant value))`.

`vertical-align` already uses a similar pattern.


> Or will I also have to implement Servo-native layout logic for these keywords?

**Comment by glasserc** [comment]
Unlike `vertical-align`, which is a single property that has specific keywords, the `MinLength` type applies to both `min-width` and `min-height`. So I couldn't apply the `vertical-align` approach directly without having two copies of the set, copy and clone functions. Instead I tried to implement `GeckoStyleCoordConvertible` for `MinLength` and put it in the `predefined_types` dict the way you explicitly told me not to :). I've opened a PR to show what I've been able to do so far (#14432).
